torch/csrc/api/src/nn/modules/transformer.cpp

```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
   return output;
 }
 
+
+// =======================================TransformerImpl================================
+TransformerImpl::TransformerImpl(TransformerOptions options_ ) : options(std::move(options_)){
+  reset();
+}
+
+void TransformerImpl::reset() {
+  // set up encoder
+  if (options.custom_encoder().is_empty()) {
+    LayerNorm norm(LayerNormOptions({options.d_model()}));
+    TransformerEncoder trans_encoder(TransformerEncoderOptions(
+      TransformerEncoderLayerOptions(options.d_model(), options.nhead())
+      .dim_feedforward(options.dim_feedforward())
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_encoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->encoder = AnyModule(trans_encoder);
+  }
+  else {
+    this->encoder = options.custom_encoder().clone();
+  }
+  this->register_module("encoder", this->encoder.ptr());
+
+  // set up decoder
+  if (options.custom_decoder().is_empty()) {
+
+    LayerNorm norm(LayerNormOptions({options.d_model()}));
+    TransformerDecoder trans_decoder(TransformerDecoderOptions(
+      TransformerDecoderLayerOptions(options.d_model(), options.nhead())
+      .dim_feedforward(options.dim_feedforward())
+      .dropout(options.dropout())
+      .activation(options.activation()), options.num_decoder_layers())
+        .norm(AnyModule(norm)));
+
+    this->decoder = AnyModule(trans_decoder);
+  }
+  else {
+    this->decoder = options.custom_decoder().clone();
+  }
+  this->register_module("decoder", this->decoder.ptr());
+
+  reset_parameters();
+}
+
+
+void TransformerImpl::reset_parameters() {
+  auto parameters = this->parameters();
+  for (auto& param : parameters) {
+    if (param.dim() > 1) {
+      torch::nn::init::xavier_uniform_(param);
+    }
+  }
+}
+
+
+Tensor TransformerImpl::forward(
+  const Tensor& src,
+  const Tensor& tgt,
+  const Tensor& src_mask,
+  const Tensor& tgt_mask,
+  const Tensor& memory_mask,
+  const Tensor& src_key_padding_mask,
+  const Tensor& tgt_key_padding_mask,
+  const Tensor& memory_key_padding_mask) {
+
+  TORCH_CHECK(src.dim() == 3 && tgt.dim() == 3,
+    "src and tgt should have 3 dimensions, but got ", src.dim(), " and ", tgt.dim());
+
+  TORCH_CHECK(src.size(1) == tgt.size(1),
+    "src and tgt should have same size at dimension 1, but got ", src.size(1), " and ", tgt.size(1));
+
+  TORCH_CHECK(src.size(2) == options.d_model() && tgt.size(2) == options.d_model(),
+    "src and tgt should have a same size as d_model at dimension 2, but got ",
+    src.size(2), " and ", tgt.size(2), " while d_model is ", options.d_model());
+
+  Tensor memory = this->encoder.forward<Tensor>(src, src_mask, src_key_padding_mask);
+  Tensor output = this->decoder.forward<Tensor>(
+    tgt, memory, tgt_mask, memory_mask, tgt_key_padding_mask, memory_key_padding_mask);
+
+  return output;
+}
+
+Tensor TransformerImpl::generate_square_subsequent_mask(int64_t sz) {
```

Comment:
I'm not sure if we want to have `generate_square_subsequent_mask` func within the transformer. Although in the python version it is, it's really not part of the paper "Attention is All Your Need".

Reply:
@zhangguanheng66 do you know why `generate_square_subsequent_mask` is exposed as a part of the Python API?

Reply:
@zhangguanheng66 if I could start from scratch, I would not include it in the transformer.

Reply:
@zou3519 @zhangguanheng66 
If we know someone already use this function in python, then I think we'd better keep this function here since otherwise, this will be an unexpected bc breaking change to user.

Reply:
It's not really BC-breaking if the functionality doesn't exist in C++ yet. It sounds like we might want to deprecate `generate_square_subsequent_mask` on the Python side?

Reply:
@zhangguanheng66 
Should I deprecate this? Or I probably can move this to the util namespace.
But I need to know the following:
1.  Is this function used specifically for transformer? (If so, probably leave this here, otherwise, can put this in nn::util)
2. Is this a generic method to generate square subsequent masks?  
3. Any other place such mask will be used within or codebase?
Thanks

Reply:
This func is mainly used to generate a square masks for the transformer application. In the word language task, word in a sentence is only allowed to see the tokens before itself so we need an attention mask here. Can we leave this func in the transformer namespace but as an independent function, rather than part of the transformer model?

Reply:
@zou3519 @zhangguanheng66 
I am going to make this static method within Transformer class. People just call Transformer::generate_sequare_subsequent_mask() instead of creating an obj and invoke it from obj.